### DISCUSSION TITLE
[BUG] When resolving a merge conflict on GitHub UI, the entire base branch of pull request is merged into the head branch.

### DISCUSSION THREAD

**Discussion by MysticHLE** [community/community#52762] [Pull Requests]
### Select Topic Area

Bug

### Body

When resolving merge conflicts through the GitHub UI, the **entire** base (destination) branch (including the merge commit that is created from the conflict resolution) is merged backwards onto the head (source) branch. Context:

**Given** (commits numbered and represented chronologically): 

* Base branch (target) with commits: C1, C2
* Head branch (source) with commits: C1, C3
* Conflicting changes between C2 and C3
* Create a PR to merge head branch into base branch, with conflicts to be resolved through GitHub UI.

**Expected branch history** after conflict resolution:

* Base branch (target) with commits: C1, C2, C3, C4
  * C4 is the merge commit with the conflicts resolved
* Head branch (source) with commits: C1, C3

**Actual branch history** after conflict resolution:

* Base branch (target) with commits: C1, C2, C3, C4 
* Head branch (source) with commits: C1, C2, C3, C4
* C4 is the merge commit with the conflicts resolved

The expected behavior is such that the head (source) branch is unmodified. This said, I do realize that this is a [documented behavior with a warning about this](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/addressing-merge-conflicts/resolving-a-merge-conflict-on-github). **However, I believe that this behavior is egregious and dangerous enough to be considered a serious bug that needs to be addressed with priority for the following reasons**:

1. The behavior is completely counter-intuitive and unlike what the `git` CLI does through a `merge` command and resolving a conflict between current (base) branch and head branch. The end behavior of the conflict resolution through the GitHub UI as implemented is more accurately branch **synchronization**, and NOT branch **merge**. When I create a PR to **merge** changes from head branch into base branch, **I expect that this is the ONLY thing it does**. If I need to further **sync** these branches in the reverse direction, then it should be up to me to create another PR to do this explicitly. 
2. The UI itself does not provide enough warning about this behavior. The UI should be intuitive to use and clear about such a dangerous behavior, and the developer should NOT be expected to read the warning about this behavior deep in the feature documentation itself. While the UI does have text indicating the direction of the merge in the conflict resolution process (and also in the final generated merge commit message), these tidbits are easily missed for the unsuspecting developer who expects things to work the same way as what they are used to in the command line. 
3. The process of performing the conflict resolution in the UI and clicking on the `Commit merge` button **actually modifies the remote head branch before the PR is even closed and before anyone can even provide feedback on the conflict resolution being taken**. That is, the mere **attempt** to resolve the conflict creates side-effects in the original head remote branch. Even if the PR is cancelled at this point, this side-effect won't be reversed. This behavior is destructive and NOT apparent in the UI. 
4. This behavior violates and muddies the core definitions of what [head branches](https://docs.github.com/en/get-started/quickstart/github-glossary#head-branch) and [base branches](https://docs.github.com/en/get-started/quickstart/github-glossary#base-branch) mean in the context of a PR. Note that when there are no conflicts, these definitions are honored. However, when there are conflicts resolved through the GitHub UI, this synchronization behavior makes it so that we can't concretely say which branch is the base or head anymore. This is made worse by the auto-generated merge commit message **in the reverse direction as intended from the original PR** for the conflict resolution commit - which is applied onto both the base and head branches.
5. This behavior makes the feature only usable for basic use cases of creating a local feature branch, pushing to remote, and then merging said change into main branch via PR. It is completely un-usable in use cases involving proper versioning and release management in which multiple remote branches are to be maintained together and merged from/to with potential conflicts to resolve while preserving certain isolation.

In the interim (whether this behavior will be changed or not), can the GitHub team **please add a setting to disable this feature on the repository and branch level**? That is, if there is a conflict detected (which the UI already does today), don't allow the PR to be created.

__Multiple developers__ from __multiple teams__ of ours ran into this issue, which unknowingly impacted our production environments __multiple times__ because they expected the UI conflict resolution to behave similarly to the CLI. As it stands, we have to continuously try to educate and remind developers on our teams to avoid using this feature as it is dangerous and completely un-usable with our release process. 

Screenshots highlighting the issues discussed above (generated from [this repository](https://github.com/MysticHLE/merge-conflict-test)): 

![Screenshot 2023-04-14 at 4 50 17 PM](https://user-images.githubusercontent.com/713685/232173197-8d3aa44b-125a-4db3-8741-e89055036530.png)

![Screenshot 2023-04-14 at 4 50 40 PM](https://user-images.githubusercontent.com/713685/232172646-2b6ad3ff-5538-4d13-87fb-dc0c2917c7c5.png)

**Comment by oguzgelal** [comment]
+1

I didn't pay attention and this behaviour resulted in production bugs today. At least is there a way to disable that button?

**Comment by yuriy-lentsevich** [comment]
@MysticHLE @oguzgelal  you can try to disable "Resolve conflicts" option in GutHub UI by adding Ruleset - "Require a pull request before merging" for the head branch

**Reply by Hellmy** [reply]
then you can not properly work with feature branches and mering them together and also protect also the feature branches.
The currenty behaviour is shown but very subtle and if you are on branch merging without protection there is also no additional hint about this.
If you have a protected branch you never can commit to that without some big red warning but during a pull request this can happen without any red text...

**Comment by alek-womple** [comment]
This bug cost us several hours today.

**Comment by brenr** [comment]
Bumping this - how is this still not fixed? Fortunately it didn't affect production.

**Comment by SUMAR7** [comment]
Bumped into this today, a lot of WIP was about to go into production because of this, luckily we have an extra branch to gather all the changes that goes to production, it saved us but I was like WTH! 😠

**Comment by perryzjc** [comment]
Bro, how come this bug is still not resolved today? It's Mar 11 now, almost one year after this post

**Reply by lidiotduvillage** [reply]
wtf...

**Comment by GreedyA1** [comment]


**Reply by MysticHLE** [reply]
Did you even read the entire post before commenting? I acknowledged the fact that this is documented behavior, but the mere act of doing so does not make such an unexpected, backwards, and intrusive behavior any less than a bug from the perspective of a Git user.

**Reply by GreedyA1** [reply]
depends on Git user I guess.

**Reply by r0bly0ns** [reply]
I'm sorry but there is no way that this is intended behaviour.

If you create a PR to merge Branch2 into Branch1 and there are no conflicts - all it does is merge the changes from Branch2 into Branch1. Any changes made in Branch1 stay there and are not merged into Bracnh2.

However, if there are any conflicts during the merge, why would the behaviour suddenly change to a bi-directional merge without warning?

This is clearly a bug

**Comment by lidiotduvillage** [comment]
I don't get how this could be unadressed for 1 YEAR now. Millions of users and such a counter intuitive and risky behavior ?! Seriously ? I don't get it.

**Reply by sivriumit** [reply]
It's been another 1 whole YEAR, and issue still persists without any answer at all. Unbelievable...

**Reply by erfanpsss** [reply]
And another year

**Comment by arowland-informatics** [comment]
This is still not resolved. We lost an entire day's worth of work today because of this.

**Comment by DSThayer** [comment]
I agree with the characterization of this issue as a bug. It silently causes a potentially very destructive action.

**Comment by tomjhall** [comment]
I have also just been bitten by this bug. It's our main which HAD branch protection `Require a pull request before merging` checked was merged when I had a PR merging from `branchX` from `main` and clicked the resolve conflicts button in the UI. This is completely unexpected from me - it violates the branch protection rules.

**Comment by d-kirsanov** [comment]
Absolutely a bug in github. Luckily we noticed what it's about to do before it actually did it! Had to do conflict resolving on the command line to avoid this.

**Comment by sai-devop** [comment]
I was completely blocked today because of this behavior. While resolving conflicts in a PR from dev to dev-mt, GitHub committed the merge directly into dev even though I never merged the PR. This caused production issues and confusion

**Comment by janderson-watchbox** [comment]
Minutes after a staging->main deploy to Production today we did main->staging and main->uat to sync changes back. The main->uat had a 1 line conflict which was resolved on the web.. that took all the pending uat code and moved it into the main branch and deployed that to Production! That's bad!

**Comment by magnusdalkvist** [comment]
This can be solved with making the base branch protected. Then a popup will show when committing the conflict resolution and you will have the option to create a new branch. After merging the pull request you get the option to delete that newly created branch. 

> I'm sorry but there is no way that this is intended behaviour.
> 
> If you create a PR to merge Branch2 into Branch1 and there are no conflicts - all it does is merge the changes from Branch2 into Branch1. Any changes made in Branch1 stay there and are not merged into Bracnh2.
> 
> However, if there are any conflicts during the merge, why would the behaviour suddenly change to a bi-directional merge without warning?
> 
> This is clearly a bug

I still agree with @r0bly0ns that the behavior is very confusing and would like to see a fix or an option to turn this off in the future.

**Reply by DSThayer** [reply]
Is this new? In the past, this scenario would completely bypass branch protections.

**Reply by magnusdalkvist** [reply]
Sorry! I did some digging and I was wrong :( this safeguard only applies to the default branch. Makes it even more stupid...

**Comment by dmitriyshashkin** [comment]
Really weird behaviour. Spent lot's of time trying to understand what's going on and why. At least provide some explanation WHY this is happening.

**Comment by ransagy** [comment]
This basically makes any flow that isn't GitHub Flow absurdly complex and dangerous to use from the UI.
I get that GitHub might be biased/opinionated, but not letting us actually make decisions on our own flows is stunning.
Azure DevOps makes a temporary branch and merges it to the target which is a lot saner.
There has to be some kinda middle ground here to avoid people tainting their branches with unexpected changes.

**Comment by paul-ennemoser-jaha** [comment]
When switching from Azure DevOps to github.com this issue has us stumpeled.
How are we supposed to work with Release Branches?

E.g. Release_2025.08 -> Release_2025.11 -> main

This issue leads to merge commit from a newer Release Branch (target branch) landing into our older Release Branch (head)
when resolving merge conflicts.

Thanks for your help.

**Comment by torext** [comment]
Absolutely atrocious state of affairs; this behaviour is completely nonsensical and, as mentioned multiple times, amounts to a sudden switch in directionality of merging simply due to a presence of conflicts. Any modification/generalization of git flow or similar branching strategies cannot be implemented properly due to this.

**Comment by ZherebcovSergey** [comment]
Hit this today on a multi-branch release flow (master/stage/dev). Resolved a conflict via GitHub UI on a stage to dev PR, and ended up with the entire dev history silently pushed into stage. Had to force-push to clean it up.
This is particularly dangerous in projects where branches are merged into production independently. The UI gives zero indication that it's about to modify the source branch as a side effect.

**Reply by Chaitanya-087** [reply]
had to hard reset to old commit in source branch 
it happend when i used github ui to resolve it :(
